File: cpp/two_sum.cpp
// Two Sum.
//
// Given an array of integers and target, return indices of the two numbers
// that sum to target.
//
// Time: O(n)
// Space: O(n)

#include <iostream>
#include <unordered_map>
#include <vector>

std::vector<int> twoSum(const std::vector<int>& nums, int target) {
    std::unordered_map<int, int> seen;
    for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
        int complement = target - nums[i];
        auto it = seen.find(complement);
        if (it != seen.end()) {
            return {it->second, i};
        }
        seen[nums[i]] = i;
    }
    return {};
}

int main() {
    auto result = twoSum({2, 7, 11, 15}, 9);
    if (result.size() == 2 && result[0] == 0 && result[1] == 1) {
        std::cout << "OK\n";
    } else {
        std::cout << "FAIL\n";
        return 1;
    }
    return 0;
}

File: typescript/twoSum.ts
/**
 * Two Sum.
 *
 * Given an array of integers and a target, return the indices of the two
 * numbers such that they add up to target.
 *
 * Time: O(n)
 * Space: O(n)
 */

export function twoSum(nums: number[], target: number): [number, number] | [] {
    const seen = new Map<number, number>();
    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];
        const j = seen.get(complement);
        if (j !== undefined) {
            return [j, i];
        }
        seen.set(nums[i], i);
    }
    return [];
}

if (require.main === module) {
    const r = twoSum([2, 7, 11, 15], 9);
    if (r.length === 2 && r[0] === 0 && r[1] === 1) {
        console.log("OK");
    } else {
        console.log("FAIL");
        process.exit(1);
    }
}
